File: src/main/java/gargoyle/calendar/web/data/ThrowableIterable.java
package gargoyle.calendar.web.data;

import java.util.Iterator;

final class ThrowableIterable implements Iterable<Throwable> {
    private Throwable exception;

    public ThrowableIterable(Throwable exception) {
        this.exception = exception;
    }

    @Override
    public Iterator<Throwable> iterator() {
        return new Iterator<Throwable>() {
            @Override
            public boolean hasNext() {
                return exception != null;
            }

            @Override
            public Throwable next() {
                exception = exception.getCause();
                return exception;
            }
        };
    }
}

File: cal-web/src/main/kotlin/gargoyle/calendar/web/data/ThrowableIterable.kt
package gargoyle.calendar.web.data

internal class ThrowableIterable(private var exception: Throwable?) : Iterable<Throwable> {
    override fun iterator(): Iterator<Throwable> {
        return object : Iterator<Throwable> {
            override fun hasNext(): Boolean {
                return exception != null
            }

            override fun next(): Throwable {
                val current = exception
                exception = exception?.cause
                return current!!
            }
        }
    }
}
